### PR TITLE
Add cypress tests

### LINKED ISSUE #91: Add acceptance tests
**Is your enhancement related to a problem? Please describe.**
As a follow on from #84, let's look to get acceptance tests to cover these two scenarios while similarly working to figure out why they occasionally fail in manual testing.

**Describe the solution you'd like**
- [ ] A test that checked the autoshare box, saved the draft, unchecked the autoshare box, published the post, and verified no tweet was sent
- [ ] A test that saved a draft with the autoshare box unchecked, checked the box on the following pageload, published the post, and verified that the tweet was sent

**Designs**

n/a

**Describe alternatives you've considered**

n/a

**Additional context**

n/a

### PR DESCRIPTION
**Pull request by thrijith**
### Description of the Change

- Setup cypress automated testing

Closes #91 

### Alternate Designs

N/A

### Possible Drawbacks

N/A

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [ x My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

### Credits

Props @

### CODE REVIEW

**Comment by jeffpaul** [comment]
@thrijith what else is needed here to bring out of draft?  (also, THANKS for helping on this!)

**Comment by thrijith** [comment]
> @thrijith what else is needed here to bring out of draft? (also, THANKS for helping on this!)

Hi @jeffpaul, the last test ran successfully here https://github.com/10up/autoshare-for-twitter/runs/5512545638?check_suite_focus=true#step:8:155 but I am seeing issue with same test when run against different WP version, don't think it is due version but something else, I'll check and see how it goes and update the PR when ready, hope that works 🙂

**Review by cadic — DISMISSED** [review]
reviewed by accident

**Comment by thrijith** [comment]
Hi @jeffpaul 

I have the test ready, but there are issues when it is running on GH actions, and the videos generated as part of the process doesn't help in pinpointing the issue here, would be really helpful if someone can take a look at this as well.

- Initially the test were [failing](https://github.com/10up/autoshare-for-twitter/actions/runs/2004947992) for all WP versions, but one passed when I ran it [individually](https://github.com/10up/autoshare-for-twitter/runs/5629860786?check_suite_focus=true).

https://github.com/10up/autoshare-for-twitter/blob/feature/add-cypress-tests/.github/workflows/test.yml#L64-L68

- The [minimum](https://github.com/10up/autoshare-for-twitter/runs/5629860895?check_suite_focus=true) and [trunk](https://github.com/10up/autoshare-for-twitter/runs/5630920530?check_suite_focus=true) versions fail too and the videos generated in test don't have info adding links to the results https://drive.google.com/drive/folders/1WG8SnhQglIgLHvwETy1DBm9BG6QywNi6?usp=sharing

- I ran those test locally and they pass so not sure how to proceed on this one right now, test on [WP minimum](https://cln.sh/7Ql1TF), test on [WP trunk](https://cln.sh/xHaKmt)

**Comment by iamdharmesh** [comment]
Hi @thrijith, 

Test failure in `minimum` and `trunk` env issue is fixed now. You may continue now on this to make it ready for review.

Thanks

**Comment by thrijith** [comment]
> Test failure in `minimum` and `trunk` env issue is fixed now. You may continue now on this to make it ready for review.

@iamdharmesh awesome, thanks! also what was the issue though? I see you moved permalink setup to be done after env setup was that is?

I'll check the requirements in https://github.com/10up/autoshare-for-twitter/issues/91and see if anything new needs to be done here or not in sometime.

**Comment by iamdharmesh** [comment]
> @iamdharmesh awesome, thanks! also what was the issue though? I see you moved permalink setup to be done after env setup was that is?

@thrijith The issue is related to some file permissions when we specify custom WordPress version for wp-env in .wp-env.json or .wp-env-override.json. In this case .htaccess file couldn’t be created in the root folder of the testing website. This is a known issue and fixed by @cadic. So, I applied his solution here

**Comment by iamdharmesh** [comment]
@thrijith considering your comment here https://github.com/10up/autoshare-for-twitter/issues/91#issuecomment-1082345137, making this PR Ready for review.

**Review by dinhtungdu — COMMENTED** [review]
@thrijith overall, the PR looks good to me, great work 🎉 !

I notice you use `cy.wait` quite often, can you try to reduce or eliminate the usage of timeout functions? Because using timeout is a signal of potential flaky tests.

**Comment by iamdharmesh** [comment]
Hi @dinhtungdu,

I have updated PR for utilizing our utils lib **@10up/cypress-wp-utils** and improved tests by removing wait and timeout-related stuff.

Please help to re-review it.

Thanks
cc: @thrijith

File: tests/cypress/support/commands.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { getRandomText } from "../support/functions";
+
+Cypress.Commands.add( 'visitAdminPage', ( page = 'index.php' ) => {
+	cy.login();
+	if ( page.includes( 'http' ) ) {
+		cy.visit( page );
+	} else {
+		cy.visit( `/wp-admin/${ page.replace( /^\/|\/$/g, '' ) }` );
+	}
+});
```
Comment (dinhtungdu) [inline tests/cypress/support/commands.js]: We have a conversation [here](https://github.com/10up/cypress-wp-utils/pull/1) about not adding the `visitAdminPage` command. I think we should follow the conclusion there in our plugin as well. What do you think? Also cc @cadic.

Comment (iamdharmesh) [reply tests/cypress/support/commands.js]: Done.

File: tests/cypress/integration/block-editor.test.js [hunk 2]
```diff
@@ -8,16 +8,20 @@ describe('Test Autoshare for Twitter with Block Editor.', () => {
 		});
 	});
 
+	beforeEach(() => {
+		cy.login();
+	});
```
Comment (dinhtungdu) [inline tests/cypress/integration/block-editor.test.js]: What do you think about applying [this technique](https://github.com/10up/cypress-wp-utils/pull/46) to improve the testing performance?

**Review by dinhtungdu — DISMISSED** [review]
LGTM! Left an optional comment but pre-approving this! Thanks so much for the great work on this PR!

File: tests/cypress/integration/block-editor.test.js (hunk 2 above)
Comment (iamdharmesh) [reply tests/cypress/integration/block-editor.test.js]: Yes, it would be great to apply it. but it's not merged yet.

Comment (iamdharmesh) [reply tests/cypress/integration/block-editor.test.js]: Ohh. you are asking to use that technique directly here on this plugin. Yes, let me do that. I was waiting for get it merged on our utils lib. :D

Comment (dinhtungdu) [reply tests/cypress/integration/block-editor.test.js]: I think we just need to copy [these lines](https://github.com/10up/cypress-wp-utils/pull/46/files#diff-6bd80aef2afe0d5b3dcd8c0f268a82b3d0b197ef23fab87ee54145dd12efe09a) and update the tests.

Comment (dinhtungdu) [reply tests/cypress/integration/block-editor.test.js]: I did [the same ](https://github.com/10up/simple-podcasting/pull/162/files#diff-6bd80aef2afe0d5b3dcd8c0f268a82b3d0b197ef23fab87ee54145dd12efe09a)for simple-podcasting.

**Review by dinhtungdu — APPROVED** [review]

**Pull request merged**